receiver/nsxreceiver/client.go

```diff
@@ -0,0 +1,236 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nsxreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/nsxreceiver"
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
+
+	dm "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/nsxreceiver/internal/model"
+)
+
+var _ (Client) = &nsxClient{}
+
+// Client is a way of interacting with the NSX REST API
+type Client interface {
+	TransportNodes(ctx context.Context) ([]dm.TransportNode, error)
+	ClusterNodes(ctx context.Context) ([]dm.ClusterNode, error)
+	NodeStatus(ctx context.Context, nodeID string, class nodeClass) (*dm.NodeStatus, error)
+	Interfaces(ctx context.Context, nodeID string, class nodeClass) ([]dm.NetworkInterface, error)
+	InterfaceStatus(ctx context.Context, nodeID, interfaceID string, class nodeClass) (*dm.NetworkInterfaceStats, error)
+}
+
+type nsxClient struct {
+	config   *Config
+	client   *http.Client
+	endpoint *url.URL
+	logger   *zap.Logger
+}
+
+var (
+	errUnauthorized = errors.New("STATUS 403, unauthorized")
+)
+
+func newClient(c *Config, settings component.TelemetrySettings, host component.Host, logger *zap.Logger) (*nsxClient, error) {
+	client, err := c.HTTPClientSettings.ToClient(host.GetExtensions(), settings)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint, err := url.Parse(c.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	return &nsxClient{
+		config:   c,
+		client:   client,
+		endpoint: endpoint,
+		logger:   logger,
+	}, nil
+}
+
+func (c *nsxClient) TransportNodes(ctx context.Context) ([]dm.TransportNode, error) {
+	body, err := c.doRequest(
+		ctx,
+		"/api/v1/transport-nodes",
+		withDefaultHeaders(),
```

Comment:
Do we ever use anything other than default headers? If not, just make this part of `doRequest`